roachtest: update mixed-version backup to use new framework

renatolabs opened the pull request:
This updates the `backup/mixed-version` roachtest to use the recently
introduced mixed-version roachtest framework (`mixedversion` package).

The main behavior exercised remains the same: backups are taken in
mixed-binary state, and those backups are restored and verified at the
end of the test. However, this commit also improves the coverage of
mixed-version backup testing in a few ways:

* **Randomization**. By virtue of using the new framework, most runs
will be different from one another since the order of actions taken by
the test will be different. Previously, backups would always be taken
with 2 nodes in the old version and 2 nodes in the new version. Now,
backups can be taken when an arbitrary number of nodes is running the
new version. As a consequence, it's also possible that some executions
will attempt backups when all nodes are running a new binary version,
but the cluster version itself has not been updated. Other points of
new randomization include the choice of the node's external dir where
backups are stored, which node to connect to when running certain
statements, and how much to wait between backups.

* **Backup Options**. Backups will randomly be created with
`revision_history` enabled, or with an `encryption_passphrase`.

* **Downgrades**. The cluster is also downgraded in mixed-version
tests. No downgrades happened in that test before this commit.

* **Workload**. Instead of using fixed call to `generate_series` to
generate data between backups, the test now runs the `bank` workload
continuously during the test. A random wait between backups allows the
workload to make changes to the underlying table during the test and
for the backups to be taken while writes are taking place.

* **Finalization**: the test _may_ attempt to create a backup as the
upgrade is finalizing (i.e., migrations are running and cluster
version is advancing).

In addition, this test will also see improved coverage as we make more
improvements to test plans generated by the `mixedversion` package.
These changes will create more backup scenarios in the future without
requiring any code changes to this test.

This test has already helped us uncover one backup bug (#97953).

Epic: CRDB-19321

Release note: None

Commit:
roachtest: update mixed-version backup to use new framework

This updates the `backup/mixed-version` roachtest to use the recently
introduced mixed-version roachtest framework (`mixedversion` package).

The main behavior exercised remains the same: backups are taken in
mixed-binary state, and those backups are restored and verified at the
end of the test. However, this commit also improves the coverage of
mixed-version backup testing in a few ways:

* **Randomization**. By virtue of using the new framework, most runs
will be different from one another since the order of actions taken by
the test will be different. Previously, backups would always be taken
with 2 nodes in the old version and 2 nodes in the new version. Now,
backups can be taken when an arbitrary number of nodes is running the
new version. As a consequence, it's also possible that some executions
will attempt backups when all nodes are running a new binary version,
but the cluster version itself has not been updated. Other points of
new randomization include the choice of the node's external dir where
backups are stored, which node to connect to when running certain
statements, and how much to wait between backups.

* **Backup Options**. Backups will randomly be created with
`revision_history` enabled, or with an `encryption_passphrase`.

* **Downgrades**. The cluster is also downgraded in mixed-version
tests. No downgrades happened in that test before this commit.

* **Workload**. Instead of using fixed call to `generate_series` to
generate data between backups, the test now runs the `bank` workload
continuously during the test. A random wait between backups allows the
workload to make changes to the underlying table during the test and
for the backups to be taken while writes are taking place.

* **Finalization**: the test _may_ attempt to create a backup as the
upgrade is finalizing (i.e., migrations are running and cluster
version is advancing).

In addition, this test will also see improved coverage as we make more
improvements to test plans generated by the `mixedversion` package.
These changes will create more backup scenarios in the future without
requiring any code changes to this test.

This test has already helped us uncover one backup bug (#97953).

Epic: CRDB-19321

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96991)
<!-- Reviewable:end -->

renatolabs commented:
Update:

1. ~~https://github.com/cockroachdb/cockroach/pull/96989~~ (merged!)
2. https://github.com/cockroachdb/cockroach/pull/96990 
3. https://github.com/cockroachdb/cockroach/pull/96991 **(this one!)**

To be reviewed: **still just the last commit**.

srosenberg reviewed (commented):
> Reviewed 3 of 9 files at r2, 10 of 10 files at r5, 5 of 5 files at r6, 2 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/backup.go` line 153 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPz5__sAwIrT07AaLo7:-NPz5__sAwIrT07AaLo8:b-qmrh87) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/backup.go#L153)):*
> ```Go
> 	csvPort := 8081
> 	csvCmd := importBankCSVServerCommand("./cockroach", csvPort)
> 	c.Run(ctx, c.All(), csvCmd+` &> logs/workload-csv-server.log < /dev/null &`)
> ```

This seems flaky. If the ssh session dies for any reason, the CSV server would be killed. I think we either need to `nohup` and `disown` or use `systemd-run`, e.g.,

```
sudo systemd-run -p LimitNOFILE=65535 --unit workload-csv-server -- $csvCmd
```
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 46 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPz8ed4DTYTT6MbtaKS:-NPz8ed4DTYTT6MbtaKT:b3jzts4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L46)):*
> ```Go
> 
> var (
> 	invalidVersionRE = regexp.MustCompile(`[^a-zA-Z0-9.]`)
> ```

Technically it's correct, but `\.` is more consistent.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 58 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPz91hV7pBY7HPZbVRn:-NPz91hV7pBY7HPZbVRo:b7nem3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L58)):*
> ```Go
> )
> 
> func sanitizeVersion(v string) string {
> ```

Not immediately clear why version needs to be sanitized. Maybe rename to `sanitizeVersionForBackup`.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 242 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzDOs3AimML-4QohxM:-NPzDOs3AimML-4QohxN:be3yvx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L242)):*
> ```Go
> 		return fmt.Errorf("error while running csv server: %w", err)
> 	})
> 	time.Sleep(time.Second) // wait for csv server to open listener
> ```

What happens if it doesn't start in time? E.g., let's say it's a scheduling chaos.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzFplTDkpWIBKVLWsz:-NPzFplTDkpWIBKVLWt-:b-pd36n9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L588)):*
> ```Go
> // in mixed-binary state, four backup collections are created: (see
> // four cases in the function body). If all nodes are running the next
> // version (which is different from the cluster version), then a
> ```

Not entirely following why the next version must be different from the cluster version when a cluster is finalized?
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 727 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzGm4mCzz1dbb3uvYh:-NPzGm4mCzz1dbb3uvYi:bhbfz3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L727)):*
> ```Go
> 			mvt := mixedversion.NewTest(ctx, t, t.L(), c, roachNodes)
> 			mvb := newMixedVersionBackup(ctx, c, roachNodes, "bank.bank")
> 			mvt.OnStartup("set short job interval", mvb.setShortJobIntervals)
> ```

This seems like an implementation detail that should perhaps be hidden from the (framework) end-user? 
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 732 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzH2GI65uZ3XzBES-q:-NPzH2GI65uZ3XzBES-r:btwa1zk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L732)):*
> ```Go
> 
> 			mvt.InMixedVersion("plan and run backups", mvb.planAndRunBackups)
> 			mvt.AfterUpgradeFinalized("verify backups", mvb.verifyBackups)
> ```

Might be worth a comment to explain why backups cannot be verified _before_ finalization. It's because we don't support restores during the mixed-version state, right?


<!-- Sent from Reviewable.io -->

smg260 reviewed (dismissed):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/backup.go` line 108 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ1Wbv2DbSQBgcG-I6a:-NQ1Wbv2DbSQBgcG-I6b:b-umiuhs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/backup.go#L108)):*
> ```Go
> // shared across the `backup/mixed-version` and
> // `declarative_schema_changer/job-compatibility-mixed-version`. Delete
> // this duplicated code once both tests have been migrated to the new
> ```

May better be documented as a TODO://
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 119 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ2-9vg04TqFOkKDTeH:-NQ2-9vg04TqFOkKDTeI:b-p3515v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L119)):*
> ```Go
> }
> 
> func newEncryptionPassphrase(rng *rand.Rand) encryptionPassphrase {
> ```

doesn't look like `rng` is used
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 200 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ2-NIv4nclO2_yiJme:-NQ2-NIv4nclO2_yiJmf:bgbhvrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L200)):*
> ```Go
> // to be observed faster, and the test to run quicker.
> func (*mixedVersionBackup) setShortJobIntervals(
> 	l *logger.Logger, rng *rand.Rand, h *mixedversion.Helper,
> ```

*nit* `l` is unused
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 354 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ1zIdYD8tNZoyEWZDd:-NQ1zIdYD8tNZoyEWZDe:b-eagwcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L354)):*
> ```Go
> func (mvb *mixedVersionBackup) saveFingerprint(
> 	l *logger.Logger,
> 	rng *rand.Rand,
> ```

*nit* Some functions use `prng`, some like this use `rng`
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 559 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ2-Vw67EN8kKRJiviM:-NQ2-Vw67EN8kKRJiviN:b7s34be) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L559)):*
> ```Go
> // enableJobAdoption (re-)enables job adoption on the given nodes.
> func (mvb *mixedVersionBackup) enableJobAdoption(
> 	l *logger.Logger, nodes option.NodeListOption, h *mixedversion.Helper,
> ```

*nit* `h` is unused


<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ5orJY4FRYF_mevGcX:-NQ5orJY4FRYF_mevGcY:b-y9oywu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L125)):*
> ```Go
> 	pass := make([]byte, passLen)
> 	for i := range pass {
> 		pass[i] = alphabet[rand.Intn(len(alphabet))]
> ```

`rng` was most likely intended to be used here.


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/backup.go` line 108 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ1Wbv2DbSQBgcG-I6a:-NQ6siO08v1ekzi8aYrV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/backup.go#L108)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

May better be documented as a TODO://
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/backup.go` line 153 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPz5__sAwIrT07AaLo7:-NQ6sluJELZLIhUkiUko:b3osau) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/backup.go#L153)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

This seems flaky. If the ssh session dies for any reason, the CSV server would be killed. I think we either need to `nohup` and `disown` or use `systemd-run`, e.g.,

```
sudo systemd-run -p LimitNOFILE=65535 --unit workload-csv-server -- $csvCmd
```
</blockquote></details>

I'm leaning towards leaving this as-is. We have lots of roachtests that run `workload` throughout the entire duration of the test, and I can't remember instances of the test flaking because the session died. Surprisingly, it seems most of our SSH-related problems come from short-lived processes/commands. I'm willing to come back to this if we see a change, but my experience so far, having run this test dozens of times at this point, is that this is not a source of flakiness (and hasn't been before these changes, as this code was already here).
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 46 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPz8ed4DTYTT6MbtaKS:-NQ6tNXq-SWiMT2-zKjt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L46)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Technically it's correct, but `\.` is more consistent.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 58 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPz91hV7pBY7HPZbVRn:-NQ6tOcb9l3NK6bWpCYR:b-hyait6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L58)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Not immediately clear why version needs to be sanitized. Maybe rename to `sanitizeVersionForBackup`.
</blockquote></details>

Function renamed, comment added.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 119 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ2-9vg04TqFOkKDTeH:-NQ6tkt9BUPRZEZMCc6U:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L119)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

doesn't look like `rng` is used
</blockquote></details>

Fixed.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ5orJY4FRYF_mevGcX:-NQ6tWKG1XLOK6ff94Ww:bf6shm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L125)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

`rng` was most likely intended to be used here.
</blockquote></details>

Heh, this comment arrived a little too late, I had already fixed it locally. I spent at least 30 minutes trying to find out why two runs with the same seed didn't produce the exact same test, and this was the reason. We definitely need a "unused arg" linter :sweat_smile: 
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 200 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ2-NIv4nclO2_yiJme:-NQ6tnMEEzzNAdmf77OD:b2zweta) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L200)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

*nit* `l` is unused
</blockquote></details>

This one is here because the function needs to conform to the `userFunc` interface.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 242 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzDOs3AimML-4QohxM:-NQ6u3cP28316CWIZ2l8:b-4cic3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L242)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

What happens if it doesn't start in time? E.g., let's say it's a scheduling chaos.
</blockquote></details>

Even though I also haven't seen this fail before, I changed it to something that should be more reliable (or at least fail with a clear error message on timeout).
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 354 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ1zIdYD8tNZoyEWZDd:-NQ6uH2pBPwUZyDT4oJ0:b-n3y44i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L354)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

*nit* Some functions use `prng`, some like this use `rng`
</blockquote></details>

Yeah, I did use `prng` on the `mixedversion` package. I'll make a note to self to be consistent when I make changes to `mixedversion` next but I don't want to make those renames in this PR.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 559 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ2-Vw67EN8kKRJiviM:-NQ6uUNQBTmXGgMVpEpb:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L559)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

*nit* `h` is unused
</blockquote></details>

Removed.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzFplTDkpWIBKVLWsz:-NQ6uWnhECEkejNxVTdf:b-mu1pd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L588)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Not entirely following why the next version must be different from the cluster version when a cluster is finalized?
</blockquote></details>

The cluster should never be finalized when this function is called. What the comment is saying is that the function may be called in mixed-binary state (some nodes run binary vN, others run vN-1); or when all nodes run the same binary version but that may not be the same as the cluster version (because the upgrade is definitely not finalized at this point).
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 727 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzGm4mCzz1dbb3uvYh:-NQ6uytD1jqBkxVFtMDC:b-6ozjkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L727)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

This seems like an implementation detail that should perhaps be hidden from the (framework) end-user? 
</blockquote></details>

I think it does belong to this test. It's fiddling with a setting for the purposes of making this test run in a reasonable amount of time, so it shouldn't be part of the framework itself. But maybe I misunderstand what you're suggesting, let me know.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 732 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzH2GI65uZ3XzBES-q:-NQ6vCmw1tu_TS_14ouE:b-h2378p) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L732)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Might be worth a comment to explain why backups cannot be verified _before_ finalization. It's because we don't support restores during the mixed-version state, right?
</blockquote></details>

Eh, so technically only _cluster_ restores are not allowed in mixed-version state so theoretically we could be running these restores in mixed-version. It's an arguable feature IMO and there's been discussions of whether we should continue supporting it.

In any case, this is in the list of things I'm planning to address in an upcoming PR (4th and hopefully last in this series).


<!-- Sent from Reviewable.io -->

renatolabs commented:
Update:

* ~https://github.com/cockroachdb/cockroach/pull/96989~ (merged!)
* ~https://github.com/cockroachdb/cockroach/pull/96990~ (merged!)
* https://github.com/cockroachdb/cockroach/pull/96991 **(this one!)**

To be reviewed: **all changes in the diff**.

srosenberg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @smg260)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ5orJY4FRYF_mevGcX:-NQ7oq8N6AoMKc6kpGew:byu5z33) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L125)):*
> We definitely need a "unused arg" linter

I forget... why is `staticcheck` not flagging it?


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ5orJY4FRYF_mevGcX:-NQBSfxCBSS4cff-9qcD:b-g97mgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L125)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

> We definitely need a "unused arg" linter

I forget... why is `staticcheck` not flagging it?
</blockquote></details>

Turns out `staticcheck` never had that functionality; I had this idea that at some point we were flagging unused args, but I guess we weren't. I imagine it's hard to reduce noise in such a linter due to how interfaces work in Go. There's some discussion in https://github.com/dominikh/go-tools/issues/124.


<!-- Sent from Reviewable.io -->

smg260 reviewed (commented):
> Reviewed 4 of 9 files at r2, 1 of 10 files at r5, 5 of 5 files at r6, 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 200 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ2-NIv4nclO2_yiJme:-NQBSs5h0vim2Z349RPR:b6kv353) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L200)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This one is here because the function needs to conform to the `userFunc` interface.
</blockquote></details>

ah! sorry i should have navigated to the interface - hard to do in reviewable 


<!-- Sent from Reviewable.io -->

renatolabs commented:
Fixed conflicts with master.

srosenberg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Great work! The new spec. for the mixed-version backup is more succinct and intuitive. It also covers more ground as evidenced by  a newly found issue [1].

[1] https://github.com/cockroachdb/cockroach/issues/97953

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-:-NQVkqPu8rnXsUI4mm0f:b-ocz3qg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @smg260)

___
*[`pkg/cmd/roachtest/tests/backup.go` line 153 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPz5__sAwIrT07AaLo7:-NQVgWpKAH92nmBl5Cz_:b-egjlzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/backup.go#L153)):*
> I can't remember instances of the test flaking because the session died.

Right, we would have seen it, otherwise. Maybe the session lasts long enough for the import to complete? I'll take a look via roachprod.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQ5orJY4FRYF_mevGcX:-NQVhbZv53j53Kc0-LWr:bqa31ii) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L125)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Turns out `staticcheck` never had that functionality; I had this idea that at some point we were flagging unused args, but I guess we weren't. I imagine it's hard to reduce noise in such a linter due to how interfaces work in Go. There's some discussion in https://github.com/dominikh/go-tools/issues/124.
</blockquote></details>

Thanks for the reference; will take a look. It might be worth a quick`semgrep` investigation to see how many false positives would arise.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 242 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzDOs3AimML-4QohxM:-NQViCle7RtmAJq01EiC:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L242)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Even though I also haven't seen this fail before, I changed it to something that should be more reliable (or at least fail with a clear error message on timeout).
</blockquote></details>

:thumbsup: 
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzFplTDkpWIBKVLWsz:-NQVifoB6TnYEk5qx3uj:biwnywk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L588)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

The cluster should never be finalized when this function is called. What the comment is saying is that the function may be called in mixed-binary state (some nodes run binary vN, others run vN-1); or when all nodes run the same binary version but that may not be the same as the cluster version (because the upgrade is definitely not finalized at this point).
</blockquote></details>

I suppose I must have missed the important part of the comment, namely `InMixedVersion`. It would be nice to assert on that state. More generally, there will be other functions which are intended to be executed only in mixed or final state. We can't really rely on comments to prevent the user error.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 727 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzGm4mCzz1dbb3uvYh:-NQVjywJ-_A6uBJdjE4J:bm7b99d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L727)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I think it does belong to this test. It's fiddling with a setting for the purposes of making this test run in a reasonable amount of time, so it shouldn't be part of the framework itself. But maybe I misunderstand what you're suggesting, let me know.
</blockquote></details>

Right, I am not actually sure what I had in mind at the time of the above comment. It does appear like a bit of fiddling, but it's specific to this test, so I don't see how this could be hidden from the user. Maybe there is a way to simplify this for other backup-related tests by moving the setup steps into a helper func.


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
> Reviewed 1 of 4 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @smg260, and @srosenberg)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NQWKqZF8m9fCdFoaYA0:-NQWKqZF8m9fCdFoaYA1:b-dp1gwu) (no related file):*
Fixed conflicts with master again.

___
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzFplTDkpWIBKVLWsz:-NQWIVZi5AspvpT7a7k7:bo29yap) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L588)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

I suppose I must have missed the important part of the comment, namely `InMixedVersion`. It would be nice to assert on that state. More generally, there will be other functions which are intended to be executed only in mixed or final state. We can't really rely on comments to prevent the user error.
</blockquote></details>

Can you explain a little further the kind of assertions you are suggesting? I'm still unsure what you are referring to when you say we are relying on comments to avoid user-error. 

To provide a little more context, as I think this is related, we tend to use the term `mixed-version` to mean different things: mixed-binary version, and unfinalized clusters (binary version != cluster version). The closure passed to `InMixedVersion` can be in either of those states. That was an intentional decision, as it's beneficial for us to be able to make the same assertions in both cases.

I've considered exposing an API that would allow these closures to specify they only want to run in one case or the other, but I've been avoiding it for as long as I can to reduce the chances of us unintentionally not covering both situations.
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 727 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzGm4mCzz1dbb3uvYh:-NQWKc3g8g-bf62IVvDo:bkiuw2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L727)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Right, I am not actually sure what I had in mind at the time of the above comment. It does appear like a bit of fiddling, but it's specific to this test, so I don't see how this could be hidden from the user. Maybe there is a way to simplify this for other backup-related tests by moving the setup steps into a helper func.
</blockquote></details>

The `setShortJobIntervals` function is already sharing the bulk of its logic with the standard `backup` roachtest. The only part that is different is how we connect to the nodes.


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzFplTDkpWIBKVLWsz:-NQWXoB_BKAnQ87OE89d:b-k5a4e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L588)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Can you explain a little further the kind of assertions you are suggesting? I'm still unsure what you are referring to when you say we are relying on comments to avoid user-error. 

To provide a little more context, as I think this is related, we tend to use the term `mixed-version` to mean different things: mixed-binary version, and unfinalized clusters (binary version != cluster version). The closure passed to `InMixedVersion` can be in either of those states. That was an intentional decision, as it's beneficial for us to be able to make the same assertions in both cases.

I've considered exposing an API that would allow these closures to specify they only want to run in one case or the other, but I've been avoiding it for as long as I can to reduce the chances of us unintentionally not covering both situations.
</blockquote></details>

I'm going to merge the PR, but feel free to respond here and I'm happy to address concerns in an upcoming PR.


<!-- Sent from Reviewable.io -->

renatolabs commented:
bors r=srosenberg

TFTR!

renatolabs commented:
bors r=srosenberg

renatolabs commented:
@smg260 I think I addressed all your comments so I took the liberty of removing your `request changes` status on this PR in order to merge it. If you have any other comments, I'm happy to address them in a follow-up PR.

srosenberg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @renatolabs, and @smg260)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzFplTDkpWIBKVLWsz:-NQWYTU23ECz681PhQd4:b-so8au0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L588)):*
> Can you explain a little further the kind of assertions you are suggesting?

A runtime assertion that would fail whenever a func is being executed in the wrong state, e.g., mixed instead of initial or final.


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96991)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/96991#-NPzFplTDkpWIBKVLWsz:-NQWZIOiBfBVTNI-SQe6:b-1ct1in) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6cdc9f4b0da8328b35527b99722d26a05fe58e7/pkg/cmd/roachtest/tests/mixed_version_backup.go#L588)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

> Can you explain a little further the kind of assertions you are suggesting?

A runtime assertion that would fail whenever a func is being executed in the wrong state, e.g., mixed instead of initial or final.
</blockquote></details>

Are you envisioning a scenario where the functions created in `mixed_version_backup.go` are used in a different context, potentially one where there's no mixed-version state? That sounds unlikely to me, given that the caller would have to instantiate a `mixedVersionBackupTest` struct.


<!-- Sent from Reviewable.io -->

renatolabs commented:
bors retry
